drivers: video: controls: rename all CIDs to Linux counterpart

josuah opened the pull request:
This aims to improve compatibility between Linux and Zephyr sources in order to facilitate debugging and implementing solution that covers both system, such as protocols to control video systems.

The CIDs numbers are expected to be exaactly the same to further help debugging.

The existing CIDs are kept as compatibility defines to the new V4L2-style names.

josuah committed:
drivers: video: controls: use the same classes IDs as Linux

The V4L2 and V4Z list of control IDs are slightly different. Use the
same CIDs as Linux, with the exact same numbers to facilitate debug
and co-development.

This also rename the GENERIC class into BASE as in Linux. No change
in the video CIDs names in this commit.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: controls: use the same CID numbers as Linux

Refers to the CID numbers found in the Linux Kernel so that each CID
of Zephyr matches the CIDs of Linux. The new CIDs introduced in
Zephyr can then follow the same number and naming found on Linux.

No renaming of the CID macro names is done yet.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: controls: Sort controls according to their class

Now that Control IDs have different classes, sort them to their
apropriate classes sections.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: controls: rename controls to Linux-like names

This uses Linux V4L2 controls as a reference to give names to the
CIDs. Apply the renaming down to the drivers that use them.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
I'm happy with applying `clang-format` too, but that means new controls submission will change many lines for alignment.

decsny commented:
@ngphibang please help review

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,42 +38,91 @@ extern "C" {
+ */
+
+/**
+ * @name Compatibility control IDs
```
I think this should be removed. Why do we need to keep compatibility while all can be changed at once ? especially the changes are not much as currently only the ov2640 has used controls IDs in upstream. On-going PRs can rebase on this if it is merged first

josuah replied:
I was thinking about 3rd-party projects. I do not have any myself though.
It is also a trivial fix on 3rd-party projects to migrate to the new IDs!

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,42 +38,91 @@ extern "C" {
+#define VIDEO_CTRL_CLASS_FLASH			0x009c0900 /**< Camera flash controls */
+#define VIDEO_CTRL_CLASS_JPEG			0x009d0900 /**< JPEG-compression controls */
+#define VIDEO_CTRL_CLASS_IMAGE_SOURCE		0x009e0900 /**< Image source controls */
+#define VIDEO_CTRL_CLASS_IMAGE_PROC		0x009f0900 /**< Image processing controls */
```
I think we should keep the same name as Linux, otherwise, it will cause more confusion. In Linux:

`#define V4L2_CTRL_CLASS_IMAGE_PROC 0x009f0000`
`#define V4L2_CID_IMAGE_PROC_CLASS_BASE (V4L2_CTRL_CLASS_IMAGE_PROC | 0x900)`

josuah replied:
I was unsure what was the best: reducing the complexity, or keep all the remapping as Linux has even if it carries no information in the context of Zephyr.
I will implement as you said, thank you for the feedback!

ngphibang replied:
No, I didn't say we need to keep all the remapping as Linux. Just here, we should do :

`#define VIDEO_CTRL_CLASS_IMAGE_PROC_**BASE**		0x009f0900`

no need to define V4L2_CTRL_CLASS_IMAGE_PROC though.

Because when looking at what is currently done:

`#define VIDEO_CTRL_CLASS_IMAGE_PROC		0x009f0900`

people will wonder why it is different in Linux where this macro is **0x009f0000**

josuah replied:
Thank you I missed that point! This is now updated through the commits.
One edge-case about the BASE class here though: https://github.com/zephyrproject-rtos/zephyr/pull/78802#discussion_r1783121678

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,42 +38,91 @@ extern "C" {
-#define VIDEO_CTRL_CLASS_VENDOR		0xFFFF0000	/**< Vendor-specific class controls */
+#define VIDEO_CTRL_CLASS_BASE			0x00980900 /**< Base class controls */
+#define VIDEO_CTRL_CLASS_CODEC			0x00990900 /**< Stateful codec controls */
+#define VIDEO_CTRL_CLASS_CAMERA			0x009a0900 /**< Camera class controls */
```
Same here:

```
#define V4L2_CTRL_CLASS_CAMERA		0x009a0000
#define V4L2_CID_CAMERA_CLASS_BASE	(V4L2_CTRL_CLASS_CAMERA | 0x900)
```

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,42 +38,91 @@ extern "C" {
-#define VIDEO_CTRL_CLASS_JPEG		0x00030000	/**< JPEG-compression controls */
-#define VIDEO_CTRL_CLASS_VENDOR		0xFFFF0000	/**< Vendor-specific class controls */
+#define VIDEO_CTRL_CLASS_BASE			0x00980900 /**< Base class controls */
+#define VIDEO_CTRL_CLASS_CODEC			0x00990900 /**< Stateful codec controls */
```
Same here : 

```
#define V4L2_CTRL_CLASS_CODEC		0x00990000
#define V4L2_CID_CODEC_BASE		(V4L2_CTRL_CLASS_CODEC | 0x900)
```

josuah replied:
This is now updated `VIDEO_CTRL_CLASS_CODEC_BASE`

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,42 +38,91 @@ extern "C" {
-#define VIDEO_CTRL_CLASS_MPEG		0x00020000	/**< MPEG-compression controls */
-#define VIDEO_CTRL_CLASS_JPEG		0x00030000	/**< JPEG-compression controls */
-#define VIDEO_CTRL_CLASS_VENDOR		0xFFFF0000	/**< Vendor-specific class controls */
+#define VIDEO_CTRL_CLASS_BASE			0x00980900 /**< Base class controls */
```
Same here :

```
#define V4L2_CTRL_CLASS_USER	0x00980000
#define V4L2_CID_BASE			(V4L2_CTRL_CLASS_USER | 0x900)
```

josuah replied:
The "base" class already did contain the `BASE` keyword. Would I rename it into `VIDEO_CID_BASE`?
Along with the other `VIDEO_CTRL_CLASS_..._BASE` renamed to `VIDEO_CID_..._BASE`?
It is a quick change if that needs be.

josuah replied:
Glad that this exception got checked still!

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,42 +38,91 @@ extern "C" {
+#define VIDEO_CTRL_CLASS_BASE			0x00980900 /**< Base class controls */
+#define VIDEO_CTRL_CLASS_CODEC			0x00990900 /**< Stateful codec controls */
+#define VIDEO_CTRL_CLASS_CAMERA			0x009a0900 /**< Camera class controls */
+#define VIDEO_CTRL_CLASS_FM_TX			0x009b0900 /**< FM Modulator controls */
```
Do some sensor use it ? Should remove ?

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,42 +38,91 @@ extern "C" {
+#define VIDEO_CTRL_CLASS_JPEG			0x009d0900 /**< JPEG-compression controls */
+#define VIDEO_CTRL_CLASS_IMAGE_SOURCE		0x009e0900 /**< Image source controls */
+#define VIDEO_CTRL_CLASS_IMAGE_PROC		0x009f0900 /**< Image processing controls */
+#define VIDEO_CTRL_CLASS_DV			0x00a00900 /**< Digital Video controls */
```
From line 56 to line 62, I think we should not add class that are actually not used by any sensors, especially these classes will be rarely used. When needed, it can be added later

josuah replied:
I made an exception for the classes: provide them all as in Linux, but thinking again, I see no reason for that exception.
It makes more sense as you suggested, will remove. Thank you!

josuah replied:
Removed from 56 to 61, but keeping the vendor control class that can be used by an application.
Did you mean to remove it also? I would try use it for USB UVC which also has a vendor control section.
I can of course wait to introduce something that uses it before adding the class!

ngphibang replied:
Yes, it's ok to keep the vendor class. I just meant the classes that are not currently used by any sensor and will be rarely used in the future.

ngphibang commented on include/zephyr/drivers/video-controls.h:
It seems github does not allow to comment on the commit message ?

> The V4L2 and Z4L2 list of control IDs are slightly different.
=> s/Z4L2/V4Z

josuah replied:
Done here, thank you!
- https://github.com/zephyrproject-rtos/zephyr/pull/78802/commits/aa34c60eedbc7d73166280e6646e13f2b70a8973

ngphibang reviewed (commented):
LGTM except a minor remark on the organization of the two last commits. Can we squash them ?

josuah commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -100,26 +100,6 @@ extern "C" {
  */
 
-/**
- * @name Compatibility control IDs
```
I tried to decompose the steps as much as possible, and give a safe backup point to fall back to, but that's also very much possible to fallback to *before the PR merge*!

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -31,13 +38,21 @@ extern "C" {
-#define VIDEO_CTRL_CLASS_JPEG		0x00030000	/**< JPEG-compression controls */
-#define VIDEO_CTRL_CLASS_VENDOR		0xFFFF0000	/**< Vendor-specific class controls */
+#define VIDEO_CTRL_CLASS_BASE			0x00980900 /**< Base class controls */
+#define VIDEO_CTRL_CLASS_CODEC_BASE		0x00990900 /**< Stateful codec controls */
```
@josuah : I don't want to delay merging this PR but because you will need to rebase this PR anyways (due to https://github.com/zephyrproject-rtos/zephyr/pull/76144 merged), I take occasion to add this comment.

I think it's better to rename the control as well to be synced with Linux, for example :

- `VIDEO_CTRL_CLASS_CAMERA_BASE` to `VIDEO_CID_CAMERA_CLASS_BASE`

josuah replied:
I appreciate the feedback. A lot easier to fix it now than after it is merged!

josuah replied:
Done. Feel free to ask for more changes as much as desired.

josuah committed:
drivers: video: controls: document each control and move classes

Make controls more readable and intuitive to extend, with definition
of each control value to help disambiguate between similar controls,
such as BRIGHTNESS vs GAIN.

Move the class definition to the top of each relevant class, like it
is on Linux for the V4L2_CID_..._CLASS_BASE variables.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
I picked the opportunity to change the documentation style, making enums in the middle of the CID list hopefully less intrusive, and make it easier to insert new CIDs (i.e. no alignment change needed).

Let me know if it was better before, easy for me to revert.

ngphibang reviewed (dismissed):
LGTM, thanks.

josuah commented:
force-push: rebase
extra commit: `<zephyr/drivers/video-controls.h>` -> `<zephyr/drivers/video/controls.h>`

Since I added plenty of many small PRs for video APIs, I thought I'd take the opportunity of this open PR to group this very related change to reduce burden. Let me know if it is best to have a separate PR for the header rename.

ngphibang commented on MAINTAINERS.yml:
Why we don't move `video.h`, `display.h`, etc. but only video-controls ?

ngphibang replied:
My understanding (maybe incorrect) is that headers in the common folder <zephyr/drivers> is public and can be used by applications, drivers, etc. while folder <zephyr/drivers/video> is specific to (video) drivers.

As in Linux, we have `v4l2-controls.h` (containing mainly the CIDs, used by both user applications and drivers) placed in the common folder `include/uapi/linux`, while `v4l2-ctrls.h` (containing only things for v4l2 drivers) is placed in `include/media` specific folder.

josuah replied:
I am not sure of the rationale behind this but this is a good question.
I followed the pattern that seemed more frequent:

<details><summary><code>~/zephyrproject/zephyr$ ls -1pd include/zephyr/drivers/*</code></summary><pre>
include/zephyr/drivers/adc/
include/zephyr/drivers/adc.h
include/zephyr/drivers/auxdisplay.h
include/zephyr/drivers/bbram.h
include/zephyr/drivers/bluetooth/
include/zephyr/drivers/bluetooth.h
include/zephyr/drivers/cache.h
include/zephyr/drivers/can/
include/zephyr/drivers/can.h
include/zephyr/drivers/cellular.h
include/zephyr/drivers/charger.h
include/zephyr/drivers/clock_control/
include/zephyr/drivers/clock_control.h
include/zephyr/drivers/comparator/
include/zephyr/drivers/comparator.h
include/zephyr/drivers/console/
include/zephyr/drivers/coredump.h
include/zephyr/drivers/counter.h
include/zephyr/drivers/dac.h
include/zephyr/drivers/dai.h
include/zephyr/drivers/disk.h
include/zephyr/drivers/display.h
include/zephyr/drivers/dma/
include/zephyr/drivers/dma.h
include/zephyr/drivers/edac.h
include/zephyr/drivers/eeprom/
include/zephyr/drivers/eeprom.h
include/zephyr/drivers/emul.h
include/zephyr/drivers/emul_bbram.h
include/zephyr/drivers/emul_fuel_gauge.h
include/zephyr/drivers/emul_sensor.h
include/zephyr/drivers/emul_stub_device.h
include/zephyr/drivers/entropy.h
include/zephyr/drivers/espi.h
include/zephyr/drivers/espi_emul.h
include/zephyr/drivers/espi_saf.h
include/zephyr/drivers/ethernet/
include/zephyr/drivers/firmware/
include/zephyr/drivers/flash/
include/zephyr/drivers/flash.h
include/zephyr/drivers/fpga.h
include/zephyr/drivers/fuel_gauge.h
include/zephyr/drivers/gnss/
include/zephyr/drivers/gnss.h
include/zephyr/drivers/gpio/
include/zephyr/drivers/gpio.h
include/zephyr/drivers/haptics/
include/zephyr/drivers/haptics.h
include/zephyr/drivers/hwinfo.h
include/zephyr/drivers/hwspinlock.h
include/zephyr/drivers/i2c/
include/zephyr/drivers/i2c.h
include/zephyr/drivers/i2c_emul.h
include/zephyr/drivers/i2s.h
include/zephyr/drivers/i3c/
include/zephyr/drivers/i3c.h
include/zephyr/drivers/ieee802154/
include/zephyr/drivers/interrupt_controller/
include/zephyr/drivers/ipm.h
include/zephyr/drivers/kscan.h
include/zephyr/drivers/led/
include/zephyr/drivers/led.h
include/zephyr/drivers/led_strip/
include/zephyr/drivers/led_strip.h
include/zephyr/drivers/loopback_disk.h
include/zephyr/drivers/lora.h
include/zephyr/drivers/mbox.h
include/zephyr/drivers/mdio.h
include/zephyr/drivers/mfd/
include/zephyr/drivers/mipi_dbi.h
include/zephyr/drivers/mipi_dsi/
include/zephyr/drivers/mipi_dsi.h
include/zephyr/drivers/misc/
include/zephyr/drivers/mm/
include/zephyr/drivers/modem/
include/zephyr/drivers/mspi/
include/zephyr/drivers/mspi.h
include/zephyr/drivers/mspi_emul.h
include/zephyr/drivers/pcie/
include/zephyr/drivers/peci.h
include/zephyr/drivers/pinctrl/
include/zephyr/drivers/pinctrl.h
include/zephyr/drivers/pm_cpu_ops/
include/zephyr/drivers/pm_cpu_ops.h
include/zephyr/drivers/power/
include/zephyr/drivers/ps2.h
include/zephyr/drivers/ptp_clock.h
include/zephyr/drivers/pwm/
include/zephyr/drivers/pwm.h
include/zephyr/drivers/regulator/
include/zephyr/drivers/regulator.h
include/zephyr/drivers/reset.h
include/zephyr/drivers/retained_mem.h
include/zephyr/drivers/rtc/
include/zephyr/drivers/rtc.h
include/zephyr/drivers/sdhc.h
include/zephyr/drivers/sensor/
include/zephyr/drivers/sensor.h
include/zephyr/drivers/sensor_attribute_types.h
include/zephyr/drivers/sensor_data_types.h
include/zephyr/drivers/serial/
include/zephyr/drivers/sip_svc/
include/zephyr/drivers/smbus.h
include/zephyr/drivers/spi/
include/zephyr/drivers/spi.h
include/zephyr/drivers/spi_emul.h
include/zephyr/drivers/stepper/
include/zephyr/drivers/stepper.h
include/zephyr/drivers/swdp.h
include/zephyr/drivers/syscon.h
include/zephyr/drivers/tee.h
include/zephyr/drivers/timer/
include/zephyr/drivers/uart/
include/zephyr/drivers/uart.h
include/zephyr/drivers/uart_emul.h
include/zephyr/drivers/uart_pipe.h
include/zephyr/drivers/usb/
include/zephyr/drivers/usb_c/
include/zephyr/drivers/video-controls.h
include/zephyr/drivers/video.h
include/zephyr/drivers/virtualization/
include/zephyr/drivers/w1.h
include/zephyr/drivers/watchdog.h
include/zephyr/drivers/wifi/
</pre></details>

Are include directories have a matching header?

```
y=0 n=0
for x in include/zephyr/drivers/*/; do [ -f "${x%/}.h" ] && y=$((y + 1)) || n=$((n + 1)); done
echo yes=$y no=$n
```

This is closer to a tie than I thought:
```
yes=26 no=18
```

However, looking at the directories that does not have a matching header:

<details><summary><code>for x in include/zephyr/drivers/*/; do if [ ! -f "${x%/}.h" ]; then echo; basename "$x"; ls "$x"; fi; done</code></summary><pre>

include/zephyr/drivers/console/:
console.h      native_posix_console.h  uart_console.h
ipm_console.h  posix_arch_console.h    uart_mcumgr.h

include/zephyr/drivers/ethernet/:
eth_adin2111.h  eth_nxp_enet.h  eth_nxp_enet_qos.h

include/zephyr/drivers/firmware/:
scmi

include/zephyr/drivers/ieee802154/:
cc1200.h

include/zephyr/drivers/interrupt_controller/:
dw_ace.h           intc_eirq_nxp_s32.h   intc_vim.h           loapic.h      sysapic.h
gd32_exti.h        intc_esp32.h          intc_wkpu_nxp_s32.h  nxp_pint.h    wuc_ite_it8xxx2.h
gic.h              intc_esp32c3.h        intc_xmc4xxx.h       riscv_clic.h
gicv3_its.h        intc_mchp_xec_ecia.h  intel_vtd.h          riscv_plic.h
gpio_intc_stm32.h  intc_ra_icu.h         ioapic.h             sam0_eic.h

include/zephyr/drivers/mfd/:
ad559x.h   axp192.h      max31790.h  npm1300.h          tle9104.h
adp5585.h  bd8lb600fs.h  nct38xx.h   nxp_lp_flexcomm.h

include/zephyr/drivers/misc/:
README     devmux  grove_lcd   pio_rpi_pico     timeaware_gpio
coresight  ft8xx   nxp_flexio  stm32_wkup_pins

include/zephyr/drivers/mm/:
mm_drv_bank.h  mm_drv_intel_adsp_mtl_tlb.h  rat.h  system_mm.h

include/zephyr/drivers/modem/:
hl7800.h  simcom-sim7080.h

include/zephyr/drivers/pcie/:
cap.h  controller.h  endpoint  msi.h  pcie.h  ptm.h  vc.h

include/zephyr/drivers/power/:
atmel_sam_supc.h

include/zephyr/drivers/serial/:
uart_altera.h    uart_async_to_irq.h  uart_intel_lw.h
uart_async_rx.h  uart_emul.h          uart_ns16550.h

include/zephyr/drivers/sip_svc/:
sip_svc_agilex_mailbox.h  sip_svc_agilex_smc.h  sip_svc_driver.h  sip_svc_proto.h

include/zephyr/drivers/timer/:
arm_arch_timer.h  nrf_grtc_timer.h  nrf_rtc_timer.h  system_timer.h  ti_dmtimer.h

include/zephyr/drivers/usb/:
emul_bc12.h  udc.h  udc_buf.h  uhc.h  usb_bc12.h  usb_dc.h

include/zephyr/drivers/usb_c/:
tcpci_priv.h  usbc_pd.h  usbc_ppc.h  usbc_tc.h  usbc_tcpc.h  usbc_vbus.h

include/zephyr/drivers/virtualization/:
ivshmem.h

include/zephyr/drivers/wifi/:
nrfwifi
</pre></details>

There is only `console/console.h` with its main header moved to the subdirectory, and all others have the pattern `<zephyr/drivers/$API.h>` where API is sometimes [`usb/udc`](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/udc/) like it is in the drivers directory.

ngphibang replied:
I just wonder if there is a reason behind (which is explained in my above comment). Perhaps @loicpoulain has some idea ?
Personnally, moving or not this header is both OK for me. Just a remark that I think the file name should be kept as `video-controls.h`. Although it is put in the `/video` folder, the filename itself should be "intuitive" on its own.

loicpoulain replied:
Personally I would not move `video-controls.h` to subdirectory, IMO there is no strong reason to have this one more than the the `video.h` one. It was cleaner to split control in a dedicated header, but it is also part of the uAPI, so it would be confusing to have two different directory levels.

josuah replied:
My understanding (maybe incorrect) is that headers in the common folder <zephyr/drivers> is public and can be used by both applications and drivers, etc. 

> while folder `<zephyr/drivers/video>` is specific to (video) drivers.

It looks like in some cases, internal APIs are defined via header files inside `~/zephyrproject/zephyr/drivers/<drivername>/<drivername>_common.h`:

<details><summary><code>~/zephyrproject/zephyr$ ls -1 drivers/*/*_common.h  drivers/*/*/*_common.h</code></summary>
<pre>
drivers/clock_control/clock_control_nrf2_common.h
drivers/clock_control/clock_stm32_ll_common.h
drivers/dma/dma_dw_common.h
drivers/i2c/i2c_nrfx_twi_common.h
drivers/i2c/i2c_nrfx_twim_common.h
drivers/lora/sx126x_common.h
drivers/lora/sx12xx_common.h
drivers/mm/mm_drv_common.h
drivers/spi/spi_litex_common.h
drivers/spi/spi_nrfx_common.h
drivers/syscon/syscon_common.h
drivers/usb/udc/udc_common.h
drivers/usb/uhc/uhc_common.h
drivers/video/video_common.h  # oops, I already started to use it on internal fork
drivers/w1/w1_ds2477_85_common.h
drivers/w1/w1_ds2482_84_common.h
drivers/watchdog/wdt_dw_common.h
</pre></details>

> Just a remark that I think the file name should be kept as video-controls.h. Although it is put in the /video folder, the filename itself should be "intuitive" on its own (?)

It looks like a `<drivername>_` prefix is frequently used:

<details><summary><code>~/zephyrproject/zephyr$ for x in include/zephyr/drivers/*/; do x=${x%/}; ls -1d $x/${x##*/}*.h; done</code></summary>
<pre>
include/zephyr/drivers/adc/adc_emul.h
include/zephyr/drivers/adc/adc_npcx_threshold.h
include/zephyr/drivers/can/can_fake.h
include/zephyr/drivers/can/can_mcan.h
include/zephyr/drivers/can/can_sja1000.h
include/zephyr/drivers/clock_control/clock_control_adsp.h
include/zephyr/drivers/clock_control/clock_control_ambiq.h
include/zephyr/drivers/clock_control/clock_control_litex.h
include/zephyr/drivers/clock_control/clock_control_numaker.h
include/zephyr/drivers/clock_control/clock_control_silabs.h
include/zephyr/drivers/console/console.h
include/zephyr/drivers/dma/dma_esp32.h
include/zephyr/drivers/dma/dma_gd32.h
include/zephyr/drivers/dma/dma_intel_lpss.h
include/zephyr/drivers/dma/dma_mcux_lpc.h
include/zephyr/drivers/dma/dma_mcux_pxp.h
include/zephyr/drivers/dma/dma_mcux_smartdma.h
include/zephyr/drivers/dma/dma_smartbond.h
include/zephyr/drivers/dma/dma_stm32.h
include/zephyr/drivers/eeprom/eeprom_fake.h
include/zephyr/drivers/flash/flash_simulator.h
include/zephyr/drivers/gnss/gnss_publish.h
include/zephyr/drivers/gpio/gpio_cmsdk_ahb.h
include/zephyr/drivers/gpio/gpio_emul.h
include/zephyr/drivers/gpio/gpio_intel.h
include/zephyr/drivers/gpio/gpio_mmio32.h
include/zephyr/drivers/gpio/gpio_nct38xx.h
include/zephyr/drivers/gpio/gpio_pcal64xxa.h
include/zephyr/drivers/gpio/gpio_sx1509b.h
include/zephyr/drivers/gpio/gpio_utils.h
include/zephyr/drivers/i3c/i3c_npcx.h
include/zephyr/drivers/mipi_dsi/mipi_dsi_mcux_2l.h
include/zephyr/drivers/mm/mm_drv_bank.h
include/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h
include/zephyr/drivers/pcie/pcie.h
include/zephyr/drivers/pinctrl/pinctrl_esp32_common.h
include/zephyr/drivers/pinctrl/pinctrl_rcar_common.h
include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h
include/zephyr/drivers/pinctrl/pinctrl_soc_sam_common.h
include/zephyr/drivers/rtc/rtc_fake.h
include/zephyr/drivers/sip_svc/sip_svc_agilex_mailbox.h
include/zephyr/drivers/sip_svc/sip_svc_agilex_smc.h
include/zephyr/drivers/sip_svc/sip_svc_driver.h
include/zephyr/drivers/sip_svc/sip_svc_proto.h
include/zephyr/drivers/stepper/stepper_fake.h
include/zephyr/drivers/stepper/stepper_trinamic.h
include/zephyr/drivers/usb/usb_bc12.h
include/zephyr/drivers/usb/usb_dc.h
</pre>
</details>

```
~/zephyrproject/zephyr$ find include/zephyr/drivers/ -name '*_*' | wc -l
186
~/zephyrproject/zephyr$ find include/zephyr/drivers/ -name '*-*'
include/zephyr/drivers/video-controls.h
include/zephyr/drivers/modem/simcom-sim7080.h
```

There is only one `-` dash and always `_` underscores used outside `video-controls.h`,
I will go with `video_controls.h` instead of `video-controls.h` if that is ok, but welcoming comments on this! i.e. if wanting to preserve the existing well-known `video-controls.h` name.

josuah replied:
> Personally I would not move `video-controls.h` to subdirectory, IMO there is no strong reason to have this one more than the the `video.h` one. It was cleaner to split control in a dedicated header, but it is also part of the uAPI, so it would be confusing to have two different directory levels.

Understood! Given this, I will keep the `video-controls.h` separate, which will help Linux users to identify the Zephyr equivalent of V4L2. Removing the last commit.

josuah commented:
force-push: CI fix in `drivers/video/capture/src/main.c` related to https://github.com/zephyrproject-rtos/zephyr/pull/79337/files#diff-fd410d45441d4e562a8cdb14320e58f2168f446e0fad0b4bb54cedd7cef29b6fR179

ngphibang commented:
Waiting for other reviewers idea about the moving of "video-controls.h", otherwise LGTM, going to approve.

kartben reviewed (changes requested):
needs a migration guide / release note entry, this will break users and they need to know what steps they have to take to update their code

josuah committed:
doc: releases: 4.1: add notes/migration guide for video CID rename

The video CIDs got changed, and users will be impacted and need to
rename some constants. Offer a strategy for finding the new names in
the migration guide for the upcoming release.

Signed-off-by: Josuah Demangeon <me@josuah.net>

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -1,85 +1,168 @@
 #include <stddef.h>
-#include <zephyr/kernel.h>
 
+#include <zephyr/device.h>
```
The `video-controls.h` only defines CID macros. It uses nothing from `device.h`, `kernel.h`, etc. Although it's not in this PR scope but I recommend to drop all included headers.

josuah replied:
Good point! Instead of shuffling these headers, I removed them.
This might have been code added and removed later (refactoring).

ngphibang commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -1,43 +1,52 @@
+#define VIDEO_CID_JPEG_CLASS_BASE		0x009d0900 /**< JPEG-compression controls */
+#define VIDEO_CID_IMAGE_SOURCE_CLASS_BASE	0x009e0900 /**< Image source controls */
+#define VIDEO_CID_IMAGE_PROC_CLASS_BASE		0x009f0900 /**< Image processing controls */
+#define VIDEO_CID_VENDOR_CLASS_BASE		0x00ff0900 /**< Vendor-specific class controls */
```
I just rechecked this `VIDEO_CID_VENDOR_CLASS_BASE`. In fact, this is Zephyr-specific and does not exist in Linux.
Its equivalent counterpart in Linux should be [V4L2_CID_PRIVATE_BASE](https://elixir.bootlin.com/linux/v6.11.1/C/ident/V4L2_CID_PRIVATE_BASE). Could you change to this one (I also need it for the upcomming control framework PR) ? Thanks !

josuah replied:
Ah right, in `include/uapi/linux/videodev2.h` I did not check there!

```c
/*  User-class control IDs defined by V4L2 */
#define V4L2_CID_MAX_CTRLS              1024
/*  IDs reserved for driver specific controls */
#define V4L2_CID_PRIVATE_BASE           0x08000000
```

I pushed the change.

ngphibang reviewed (changes requested):
VIDEO_CID_VENDOR_CLASS_BASE need to be changed to VIDEO_CID_PRIVATE_BASE to be aligned with Linux

josuah commented:
force-push:
- Update the VENDOR class to PRIVATE like in Linux.
- Rebase on latest main

kartben commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -1,85 +1,162 @@
-#define VIDEO_CTRL_CLASS_VENDOR  0xFFFF0000 /**< Vendor-specific class controls */
+#define VIDEO_CID_BASE 0x00980900
+
+/** Amount of perceived light of the image, the luma (Y') value. */
+#define VIDEO_CID_BRIGHTNESS (VIDEO_CID_BASE + 0)
+
+/** Amount of difference between the bright colors and dark colors. */
+#define VIDEO_CID_CONTRAST (VIDEO_CID_BASE + 1)
+
+/** Colorfulness of the image while preserving its brightness */
+#define VIDEO_CID_SATURATION (VIDEO_CID_BASE + 2)
+
+/** Shift in the tint of every colors, clockwise in a RGB color wheel */
+#define VIDEO_CID_HUE (VIDEO_CID_BASE + 3)
+
+/** Amount of time an image sensor is exposed to light, affecting the brightness */
+#define VIDEO_CID_EXPOSURE (VIDEO_CID_BASE + 17)
```
I appreciate the effort to make sure all public API is documented but FWIW I find these oddly worded for some reason. I guess that's what happens when the names are already mostly self-explanatory and one doesn't want to literally paraphrase them in the comment. Not blocking though :)

ngphibang replied:
Yes, I have same thoughts too and sometimes, the explanation is not totally correct as it is difficult to explain in short words, so it is better to not explain :-).

josuah replied:
Reading them again, these do not help understanding what the control do.

Better not risk defining them slightly differently than Linux ones.
If there need to be any doc effort, this can go in Linux.

New PR to address these small changes. :)

josuah replied:
How about a link to this page for reference?
https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/control.html

ngphibang replied:
Yes, I think it's a good idea but I am not sure if it's OK in Zephyr to refer to documentation of a different project.
